docs/proposals/memory-management.md

```diff
@@ -0,0 +1,185 @@
+
+1. Expose the memory capacity and usage of the node.
+
+2. Apply memory limits to user containers and individual processes on the node.
+
+3. Set up cgroups to track the memory consumption of itself and other system processes like docker daemon, kube-proxy, sshd, etc. The cgroups being setup by kubelet are,
+  a. `/docker-daemon`: The docker daemon is under this cgroup. This cgroup is limited to 70% of the total available memory on the node.
+  b. `/system`: OS processes, excepting init, are under this cgroup.
+  c. `/kubelet`: Kubelet is under this cgroup.
+  d. `/kube-proxy`: The kube-proxy is under this cgroup.
+
+4. Protect system processes from memory pressure on the node:
+   The `oom_score_adj` for kubelet, kube-proxy and other system processes is set to `-999`
+   The `oom_score_adj` for docker daemon is set to `-990`. Docker's memory usage has been unpredicatble in the past and docker daemons restarts are possible under system pressure.
+
+5. Manage Quality of service for user containers: Three classes of containers are currently supported:
+
+   a. `Guaranteed`: Memory is always guaranteed up to the request. These containers will not OOM unless they exceed their request, or system processes consume too much memory, or the kernel overhead is high and the memory in the node is being fully consumed. Note that `request` equals `limit` for this class of containers.
+
+   b. `Burstable`: Memory is guaranteed up until the usage exceeds user specified `request`. These containers do not get killed as soon as they exceed their request. Kernel memory overhead and system processes memory usage will impact memory availability for this class of containers. They are more likely to be killed by the kernel OOM killer once they exceed their request.
+
+   c. `Best-Effort`: Memory is never guaranteed. This class of containers are the first ones to be killed by the kernel whenever there is memory pressure on the node.
+
+6. Detect OOM kills on containers: Docker tracks OOM kills on its containers.
+Kubelet exposes this information via [ContainerStatus.State](../../pkg/api/v1/types.go#L1033), where the state is set to `ContainerStateTerminated` and `Reason` is set to `OOM Killed`.
+Kubelet has a built-in exponential backoff mechanism that prevents OOMing containers from entering a crash loop and bringing down the node.
+
+
+#### What are some of the known limitations?
+
+1. The scheduler does not take into account the total usage on the node. Hence, memory can be easily over-committed. Whenever a job is scheduled without resource requests, the node becomes over-committed.
+
+2. Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+
+3. A container in a pod with restart policy set to `Always` will be restarted continuously even if it OOMs. This is alleviated with the crash loop backoff feature. To improve node stability, that pod should be evicted and run on a different node.
```

Comment:
I think we should distinguish the type of OOM for clarity - system oom versus local container OOM.

If a container has request[memory] == limit[memory], and it repeatedly gets a container OOMKilled on the same node (N times with Y interval), then there is really nothing we can do except potentially permanently fail the pod with a reason as being sized too small.  We then may want to have any controller that spawns those pod's introspect to know if its frequently occurring that we should do some back-off.  If its a slow memory leak though that occurs after a large period of time, just restarting container seems fine.

Where memory request is provided or unspecified, I think there is a valid sentiment on more proactively permanently failing the pod on the node.  That could be based on the ratio between the request and limit to determine how pro-active that decision is made on the node.